### PR TITLE
add support for runtime metrics

### PR DESCRIPTION
**Pull request by RafalSkolasinski**
**What this PR does / why we need it**:

Adds support for thread/process safe runtime metrics / tags.

**Which issue(s) this PR fixes**:

Closes https://github.com/SeldonIO/seldon-core/issues/1735

**Special notes for your reviewer**:

Introduced `seldon_core.user_model.SeldonResponse` class that can be used as output of `predict` and other methods on User Models in order to set runtime metrics and tags.
For example
```python
from seldon_core.user_model import SeldonResponse

class Model:
    ...
    def predict(self, features, names=[], meta=[]):
        metrics = [{"type": "COUNTER", "key": "predictions_counter", "value": 1}]
        tags = {"a": 42, "z": 5}
        return SeldonResponse(data=features.tolist(), metrics=metrics, tags=tags)
```

**Does this PR introduce a user-facing change?**:

```release-note
Added support for runtime metrics and tags.
```

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Tue Sep 15 19:30:17 UTC 2020
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=2

**Comment by seldondev** [comment]
Tue Sep 15 19:30:29 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=1

**Comment by seldondev** [comment]
Tue Sep 15 21:24:57 UTC 2020
The logs for [pr-build] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=3

**Comment by seldondev** [comment]
Tue Sep 15 21:25:00 UTC 2020
The logs for [lint] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=4

**Comment by seldondev** [comment]
Tue Sep 15 21:28:12 UTC 2020
The logs for [integration] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=5

**Comment by seldondev** [comment]
@RafalSkolasinski: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | c9d8301213588d515033dca30f83ef0d05ba18d0 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Tue Sep 15 23:28:55 UTC 2020
The logs for [integration] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=6

File: python/seldon_core/seldon_methods.py
Comment (axsaucedo) [inline python/seldon_core/seldon_methods.py:385]: Looking at what the function is doing, it makes more sense for it to be part of the creation of the client_response above (same for all the ones below)

File: python/seldon_core/seldon_methods.py
Comment (axsaucedo) [inline python/seldon_core/seldon_methods.py:444]: Route is expected to return int so this wouldn't apply

File: python/seldon_core/user_model.py
Comment (axsaucedo) [inline python/seldon_core/user_model.py:126]: Name should change to something more specific, as this is returned in aggregate, transform_, etc

File: python/seldon_core/user_model.py
Comment (axsaucedo) [inline python/seldon_core/user_model.py:132]: Should add type to data for consistency

**Review by axsaucedo — CHANGES_REQUESTED** [review]

File: python/seldon_core/user_model.py
Comment (RafalSkolasinski) [reply python/seldon_core/user_model.py:126]: How would you suggest to name it?

Comment (RafalSkolasinski) [reply python/seldon_core/user_model.py:126]: Maybe `SeldonResponse` or `ClientResponse`.

File: python/seldon_core/seldon_methods.py
Comment (RafalSkolasinski) [reply python/seldon_core/seldon_methods.py:385]: When started working on it inside `predict` method I was initially planning to keep it as `SeldonPrediction` as long as possible but then there was a custom operations on data in `feedback` and `route` methods so maybe your suggestion is better.

File: python/seldon_core/seldon_methods.py
Comment (RafalSkolasinski) [reply python/seldon_core/seldon_methods.py:444]: I thought we'd like users to be able to still define runtime metrics / tags on route method?

Comment (RafalSkolasinski) [reply python/seldon_core/seldon_methods.py:444]: ah, right, we have
```
    def route(
        self, features: Union[np.ndarray, str, bytes], feature_names: Iterable[str]
    ) -> int:
        raise SeldonNotImplementedError("route is not implemented")
```

okay, fair enough, removing from route

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **axsaucedo**
You can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Wed Sep 16 09:47:15 UTC 2020
The logs for [pr-build] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=7

**Comment by seldondev** [comment]
Wed Sep 16 09:47:16 UTC 2020
The logs for [lint] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=8

**Comment by seldondev** [comment]
Wed Sep 16 09:48:54 UTC 2020
The logs for [integration] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=9

File: python/seldon_core/seldon_methods.py
Comment (RafalSkolasinski) [reply python/seldon_core/seldon_methods.py]: Removed runtime metrics / tags from route method

File: python/seldon_core/seldon_methods.py
Comment (RafalSkolasinski) [reply python/seldon_core/seldon_methods.py]: Movied logic into corresponding methods inside user_model.py file

File: python/seldon_core/user_model.py
Comment (RafalSkolasinski) [reply python/seldon_core/user_model.py]: Added

File: python/seldon_core/user_model.py
Comment (RafalSkolasinski) [reply python/seldon_core/user_model.py]: Went with `ClientResponse`, can still adjust easily

**Comment by seldondev** [comment]
@RafalSkolasinski: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | c58c07842bd966f2b3ea4fbb5085cf9381b6e214 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Wed Sep 16 15:52:43 UTC 2020
The logs for [pr-build] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=10

**Comment by seldondev** [comment]
Wed Sep 16 15:52:59 UTC 2020
The logs for [lint] [11] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/11.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=11

**Comment by seldondev** [comment]
Wed Sep 16 15:57:23 UTC 2020
The logs for [integration] [12] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/12.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=12

**Comment by seldondev** [comment]
Wed Sep 16 18:30:11 UTC 2020
The logs for [pr-build] [13] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/13.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=13

**Comment by seldondev** [comment]
Wed Sep 16 18:30:17 UTC 2020
The logs for [lint] [14] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/14.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=14

**Review by axsaucedo — COMMENTED** [review]
Looks good, final things as discussed:
* Update ClientResponse -> SeldonResponse
* Add nblink to example and list on notebooks.rst
* Add brief mention on this in the python module docs
* Add point on the existing issue on updating python documentation to add further insights

**Comment by seldondev** [comment]
Thu Sep 17 12:46:09 UTC 2020
The logs for [pr-build] [15] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/15.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=15

**Comment by seldondev** [comment]
Thu Sep 17 12:46:17 UTC 2020
The logs for [lint] [16] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/16.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=16

**Comment by RafalSkolasinski** [comment]
/test integration
/test notebooks

@axsaucedo ready for final review (modified slightly function that updates metrics so it also will validate if runtime one is defined properly)

**Comment by seldondev** [comment]
Thu Sep 17 17:40:34 UTC 2020
The logs for [lint] [18] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/18.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=18

**Comment by seldondev** [comment]
Thu Sep 17 17:40:45 UTC 2020
The logs for [pr-build] [17] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/17.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=17

**Comment by seldondev** [comment]
Thu Sep 17 17:41:16 UTC 2020
The logs for [notebooks] [19] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/19.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=19

**Comment by seldondev** [comment]
Thu Sep 17 17:43:18 UTC 2020
The logs for [integration] [20] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/20.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=20

**Comment by seldondev** [comment]
@RafalSkolasinski: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by axsaucedo — APPROVED** [review]

**Comment by axsaucedo** [comment]
@RafalSkolasinski it seems the lint job failed, may need rebase

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/2442#issuecomment-694407154 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/2442#issuecomment-694407154 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com#issuecomment-694407154/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
@RafalSkolasinski: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test lint`
integration | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@RafalSkolasinski: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test lint`
integration | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test integration`
notebooks | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
New changes are detected. LGTM label has been removed.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/2442#issuecomment-694407154 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Fri Sep 18 09:35:19 UTC 2020
The logs for [pr-build] [21] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/21.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=21

**Comment by seldondev** [comment]
Fri Sep 18 09:35:29 UTC 2020
The logs for [lint] [22] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/22.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=22

**Comment by seldondev** [comment]
Fri Sep 18 09:35:42 UTC 2020
The logs for [integration] [23] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/23.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=23

**Comment by RafalSkolasinski** [comment]
Just so we can track failed tests.

Integration: `test_label_update[1.0.2]` 
Note: most likely just old good flakiness

Notebooks: `TestNotebooks.test_grpc_metadat` and `TestNotebooks.test_explainer_examples`

**Comment by seldondev** [comment]
@RafalSkolasinski: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test integration`
notebooks | b7b33ea0741e13f87acfc624c1d066c3953501dc | [link]() | `/test notebooks`
lint | 6b184b784f0824fd50d332c920d550a193ab4b5d | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Fri Sep 18 09:45:48 UTC 2020
The logs for [pr-build] [24] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/24.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=24

**Comment by seldondev** [comment]
Fri Sep 18 09:45:58 UTC 2020
The logs for [lint] [25] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/25.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=25

**Comment by seldondev** [comment]
Fri Sep 18 09:46:06 UTC 2020
The logs for [integration] [26] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2442/26.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2442 --build=26

**Pull request merged**
